// zarm/es/calendar/Month.js
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _typeof from "@babel/runtime/helpers/typeof";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component, isValidElement } from 'react';
import classnames from 'classnames';
import DateTool from '../utils/date';

var CalendarMonthView = /*#__PURE__*/function (_Component) {
  _inherits(CalendarMonthView, _Component);

  var _super = _createSuper(CalendarMonthView);

  // 月份缓存数据
  // 月份最小值
  // 月份最大值
  // 上次是否落点在当前月份内
  // 当前组件是否需要更新
  // 当前月份的dom
  function CalendarMonthView(_props) {
    var _this;

    _classCallCheck(this, CalendarMonthView);

    _this = _super.call(this, _props);
    _this.min = void 0;
    _this.max = void 0;
    _this.lastIn = false;
    _this.isRefresh = true;
    _this.node = void 0;

    _this.anchor = function () {
      if (_this.node && _this.node.scrollIntoViewIfNeeded) {
        _this.node.scrollIntoViewIfNeeded();
      }
    };

    _this.checkRefresh = function (props) {
      var dateMonth = props.dateMonth,
          value = props.value,
          min = props.min,
          max = props.max,
          dateRender = props.dateRender,
          disabledDate = props.disabledDate;
      var _this$props = _this.props,
          dateRenderProp = _this$props.dateRender,
          disabledDateProp = _this$props.disabledDate;
      var dateMonthState = _this.state.dateMonth;

      if (dateRender !== dateRenderProp || disabledDate !== disabledDateProp) {
        return true;
      }

      if (+dateMonth - +dateMonthState !== 0) {
        return true;
      }

      if (+min - +_this.min !== 0 || +max - +_this.max !== 0) {
        return true;
      }

      var isIn;

      if (value.length > 0) {
        var currMonth = DateTool.cloneDate(dateMonth, 'dd', 1);
        var min1 = DateTool.cloneDate(value[0], 'dd', 1);
        var max1 = DateTool.cloneDate(value[value.length - 1], 'dd', 1);
        isIn = currMonth >= min1 && currMonth <= max1;
      }

      var result = !(!isIn && !_this.lastIn);
      _this.lastIn = isIn;
      return result;
    };

    _this.renderDay = function (day, year, month, firstDay) {
      var _className;

      var _this$props2 = _this.props,
          prefixCls = _this$props2.prefixCls,
          dateRender = _this$props2.dateRender,
          onDateClick = _this$props2.onDateClick;
      var date = new Date(year, month, day);
      var isToday = CalendarMonthView.cache.now === "".concat(year, "-").concat(month, "-").concat(day);

      var status = _this.checkStatus(date);

      var txt = date && dateRender && dateRender(date) || '';

      if (_typeof(txt) === 'object') {
        if (! /*#__PURE__*/isValidElement(txt)) {
          console.warn('dateRender函数返回数据类型错误，请返回基本数据类型或者reactNode');
          txt = '';
        }
      }

      var className = (_className = {
        d6: (day + firstDay) % 7 === 0,
        d7: (day + firstDay) % 7 === 1
      }, _defineProperty(_className, "".concat(prefixCls, "__day--disabled"), status.disabled), _defineProperty(_className, "".concat(prefixCls, "__day--today"), isToday), _defineProperty(_className, "".concat(prefixCls, "__day--selected"), status.isSelected), _defineProperty(_className, "".concat(prefixCls, "__day--range"), status.isRange), _defineProperty(_className, 'range-start', status.rangeStart), _defineProperty(_className, 'range-end', status.rangeEnd), _defineProperty(_className, "firstday-".concat(firstDay), day === 1 && firstDay), _className);
      return /*#__PURE__*/React.createElement("li", {
        key: "".concat(year, "-").concat(month, "-").concat(day),
        className: classnames("".concat(prefixCls, "__day"), className),
        onClick: function onClick() {
          return !status.disabled && date && onDateClick && onDateClick(date);
        }
      }, txt && /*#__PURE__*/React.createElement("div", {
        className: "".concat(prefixCls, "__day__content")
      }, txt) || '');
    };

    _this.renderContent = function (year, month) {
      var data = DateTool.getCurrMonthInfo(year, month);
      var firstDay = data.firstDay,
          dayCount = data.dayCount;
      return Array.from({
        length: dayCount
      }).map(function (_item, i) {
        return _this.renderDay(i + 1, year, month, firstDay);
      });
    };

    _this.min = _props.min;
    _this.max = _props.max;
    _this.state = {
      value: _props.value,
      dateMonth: _props.dateMonth
    };
    _this.checkStatus = _this.checkStatus.bind(_assertThisInitialized(_this));
    return _this;
  }

  _createClass(CalendarMonthView, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps) {
      this.isRefresh = this.checkRefresh(nextProps);

      if (this.isRefresh) {
        this.min = nextProps.min;
        this.max = nextProps.max;
      }

      return this.isRefresh;
    }
  }, {
    key: "checkStatus",
    value: // 日期状态: 选中，区间
    function checkStatus(date) {
      var _this$props3 = this.props,
          min = _this$props3.min,
          max = _this$props3.max,
          disabledDate = _this$props3.disabledDate;
      var _this$state$value = this.state.value,
          value = _this$state$value === void 0 ? [] : _this$state$value;
      var disabled = date < DateTool.cloneDate(min, 'd', 0) || date > DateTool.cloneDate(max, 'd', 0);
      var res = {
        disabled: disabled || disabledDate && disabledDate(date),
        isSelected: value.some(function (item) {
          return DateTool.isOneDay(date, item);
        }),
        isRange: value.length > 1 && date > value[0] && date < value[value.length - 1],
        rangeStart: value.length > 1 && DateTool.isOneDay(date, value[0]),
        rangeEnd: value.length > 1 && DateTool.isOneDay(date, value[value.length - 1])
      };
      this.lastIn = this.lastIn || res.isSelected || res.isRange;
      return res;
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var prefixCls = this.props.prefixCls;
      var dateMonth = this.state.dateMonth;
      var year = dateMonth.getFullYear();
      var month = dateMonth.getMonth();
      var monthKey = "".concat(year, "-").concat(month);
      return /*#__PURE__*/React.createElement("section", {
        key: monthKey,
        className: "".concat(prefixCls, "__month"),
        title: "".concat(year, "\u5E74").concat(month + 1, "\u6708"),
        ref: function ref(n) {
          _this2.node = n;
        }
      }, /*#__PURE__*/React.createElement("ul", null, this.renderContent(year, month)));
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(nextProps, state) {
      if ( // eslint-disable-next-line operator-linebreak
      'value' in nextProps && nextProps.value !== state.value || 'dateMonth' in nextProps && nextProps.dateMonth !== state.dateMonth) {
        return {
          value: nextProps.value,
          dateMonth: nextProps.dateMonth
        };
      }

      return null;
    }
  }]);

  return CalendarMonthView;
}(Component);

CalendarMonthView.displayName = 'CalendarMonthView';
CalendarMonthView.defaultProps = {
  prefixCls: 'za-calendar',
  value: [],
  dateMonth: new Date(),
  min: new Date(),
  max: new Date(),
  dateRender: function dateRender(date) {
    return date.getDate();
  },
  disabledDate: function disabledDate() {
    return false;
  }
};
CalendarMonthView.now = new Date();
CalendarMonthView.cache = {
  now: "".concat(CalendarMonthView.now.getFullYear(), "-").concat(CalendarMonthView.now.getMonth(), "-").concat(CalendarMonthView.now.getDate())
};
export { CalendarMonthView as default };
